File: script/behavior_verifier.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: 2015 - 2026 Rime community
#
# SPDX-License-Identifier: GPL-3.0-or-later
"""Verify keypress behavior in a resolved component definition.

Walks every keyboard and every action field (`click`, `long_click`, swipe_*,
`composing`). If the action value names a preset key, the preset key definition
is checked. If it does not name a preset key, it is treated as literal output.

Also checks keyboard-level references such as `ascii_keyboard` and
`landscape_keyboard`.
"""

from __future__ import annotations

from typing import Any

ACTION_FIELDS = {
    "click",
    "long_click",
    "swipe_up",
    "swipe_down",
    "swipe_left",
    "swipe_right",
    "composing",
}

SPECIAL_SELECTS = {".next", ".last", ".default", ".ascii", ".last_lock"}


def verify(sections: dict[str, Any]) -> list[str]:
    errors: list[str] = []
    preset_keys = sections.get("preset_keys", {})
    keyboards = sections.get("preset_keyboards", {})
    switch_policy = sections.get("keyboard_switch_policy", {})

    # Keyboard-level references.
    for kb_name, kb in keyboards.items():
        if not isinstance(kb, dict):
            errors.append(f"keyboard '{kb_name}' is not a mapping")
            continue
        for ref_field in ("ascii_keyboard", "landscape_keyboard"):
            ref = kb.get(ref_field)
            if isinstance(ref, str) and ref and ref not in keyboards:
                errors.append(
                    f"keyboard '{kb_name}'.{ref_field} references unknown keyboard '{ref}'"
                )

        for index, key in enumerate(kb.get("keys", []) or []):
            if not isinstance(key, dict):
                errors.append(f"keyboard '{kb_name}' key[{index}] is not a mapping")
                continue
            for field in ACTION_FIELDS:
                value = key.get(field)
                if value is None:
                    continue
                if isinstance(value, str) and value in preset_keys:
                    errors.extend(
                        verify_preset_key(
                            preset_keys[value],
                            where=f"keyboard '{kb_name}' key[{index}] {field} '{value}'",
                            keyboards=keyboards,
                        )
                    )
                # Anything else is literal output (number/string not matching a
                # preset key name) and is considered valid.

    # Keyboard switch policy references.
    for field in ("default_keyboard", "ascii_keyboard"):
        target = switch_policy.get(field)
        if isinstance(target, str) and target and target not in keyboards:
            errors.append(f"keyboard_switch_policy.{field} references unknown keyboard '{target}'")
    aux = switch_policy.get("aux", {})
    if isinstance(aux, dict):
        for role, target in aux.items():
            if isinstance(target, str) and target and target not in keyboards:
                errors.append(
                    f"keyboard_switch_policy.aux.{role} references unknown keyboard '{target}'"
                )

    # style.keyboards list references.
    style = sections.get("style", {})
    style_keyboards = style.get("keyboards") if isinstance(style, dict) else None
    if isinstance(style_keyboards, list):
        for name in style_keyboards:
            if isinstance(name, str) and name not in keyboards and name != ".default":
                errors.append(f"style.keyboards references unknown keyboard '{name}'")

    return errors


def verify_preset_key(
    definition: Any,
    where: str,
    keyboards: dict[str, Any] | None = None,
) -> list[str]:
    errors: list[str] = []
    if not isinstance(definition, dict):
        errors.append(f"{where}: preset key definition is not a mapping")
        return errors

    select = definition.get("select")
    if select is not None:
        if not isinstance(select, str) or not select:
            errors.append(f"{where}: 'select' must be a non-empty string")
        elif select not in SPECIAL_SELECTS and keyboards is not None and select not in keyboards:
            errors.append(f"{where}: 'select' references unknown keyboard '{select}'")

    toggle = definition.get("toggle")
    if toggle is not None:
        if not isinstance(toggle, str) or not toggle:
            errors.append(f"{where}: 'toggle' must be a non-empty string")
        states = definition.get("states")
        if not isinstance(states, list) or not states:
            errors.append(f"{where}: toggle '{toggle}' must have a non-empty 'states' list")

    command = definition.get("command")
    if command is not None and (not isinstance(command, str) or not command):
        errors.append(f"{where}: 'command' must be a non-empty string")

    return errors

File: app/src/main/java/com/osfans/trime/data/theme/component/BehaviorVerifier.kt
// SPDX-FileCopyrightText: 2015 - 2026 Rime community
//
// SPDX-License-Identifier: GPL-3.0-or-later

package com.osfans.trime.data.theme.component

import com.osfans.trime.util.yaml.Node
import com.osfans.trime.util.yaml.mapping
import com.osfans.trime.util.yaml.sequence
import com.osfans.trime.util.yaml.string

/**
 * Pure verification of keypress behavior in a resolved component definition.
 *
 * Walks every keyboard and every action field. If the action value names a
 * preset key, the preset key definition is checked; otherwise it is treated as
 * literal output.
 */
object BehaviorVerifier {
    private val ACTION_FIELDS =
        setOf(
            "click",
            "long_click",
            "swipe_up",
            "swipe_down",
            "swipe_left",
            "swipe_right",
            "composing",
        )

    private val SPECIAL_SELECTS =
        setOf(".next", ".last", ".default", ".ascii", ".last_lock")

    fun verify(sections: Map<String, Node.Mapping>): List<String> {
        val errors = mutableListOf<String>()
        val presetKeys = sections["preset_keys"] ?: Node.Mapping()
        val keyboards = sections["preset_keyboards"] ?: Node.Mapping()
        val switchPolicy = sections["keyboard_switch_policy"] ?: Node.Mapping()

        keyboards.pairs.forEach { (nameNode, valueNode) ->
            val name = nameNode.string ?: return@forEach
            val keyboard = valueNode.mapping ?: run {
                errors += "keyboard '$name' is not a mapping"
                return@forEach
            }
            verifyKeyboardReferences(name, keyboard, keyboards, errors)
            val keys = keyboard["keys"]?.sequence?.nodes ?: emptyList()
            keys.forEachIndexed { index, keyNode ->
                val key = keyNode.mapping ?: run {
                    errors += "keyboard '$name' key[$index] is not a mapping"
                    return@forEachIndexed
                }
                verifyKeyActions(name, index, key, presetKeys, keyboards, errors)
            }
        }

        listOf("default_keyboard", "ascii_keyboard").forEach { field ->
            val target = switchPolicy[field]?.string
            if (!target.isNullOrEmpty() && keyboards[target] == null) {
                errors += "keyboard_switch_policy.$field references unknown keyboard '$target'"
            }
        }
        switchPolicy["aux"]?.mapping?.pairs?.forEach { (role, targetNode) ->
            val target = targetNode.string
            if (!target.isNullOrEmpty() && keyboards[target] == null) {
                errors += "keyboard_switch_policy.aux.${role.string} references unknown keyboard '$target'"
            }
        }

        val style = sections["style"] ?: Node.Mapping()
        style["keyboards"]?.sequence?.nodes?.mapNotNull { it.string }?.forEach { name ->
            if (name != ".default" && keyboards[name] == null) {
                errors += "style.keyboards references unknown keyboard '$name'"
            }
        }

        return errors
    }

    private fun verifyKeyboardReferences(
        name: String,
        keyboard: Node.Mapping,
        keyboards: Node.Mapping,
        errors: MutableList<String>,
    ) {
        listOf("ascii_keyboard", "landscape_keyboard").forEach { field ->
            val target = keyboard[field]?.string
            if (!target.isNullOrEmpty() && keyboards[target] == null) {
                errors += "keyboard '$name'.$field references unknown keyboard '$target'"
            }
        }
    }

    private fun verifyKeyActions(
        keyboardName: String,
        index: Int,
        key: Node.Mapping,
        presetKeys: Node.Mapping,
        keyboards: Node.Mapping,
        errors: MutableList<String>,
    ) {
        ACTION_FIELDS.forEach { field ->
            val value = key[field] ?: return@forEach
            val text = value.string ?: return@forEach
            if (presetKeys[text]?.mapping != null) {
                verifyPresetKey(
                    presetKeys[text]!!.mapping!!,
                    where = "keyboard '$keyboardName' key[$index] $field '$text'",
                    keyboards = keyboards,
                    errors = errors,
                )
            }
        }
    }

    private fun verifyPresetKey(
        definition: Node.Mapping,
        where: String,
        keyboards: Node.Mapping,
        errors: MutableList<String>,
    ) {
        definition["select"]?.string?.let { select ->
            if (select.isEmpty()) {
                errors += "$where: 'select' must be a non-empty string"
            } else if (select !in SPECIAL_SELECTS && keyboards[select] == null) {
                errors += "$where: 'select' references unknown keyboard '$select'"
            }
        }

        definition["toggle"]?.string?.let { toggle ->
            if (toggle.isEmpty()) {
                errors += "$where: 'toggle' must be a non-empty string"
            }
            val states = definition["states"]?.sequence?.nodes
            if (states.isNullOrEmpty()) {
                errors += "$where: toggle '$toggle' must have a non-empty 'states' list"
            }
        }

        definition["command"]?.string?.let { command ->
            if (command.isEmpty()) {
                errors += "$where: 'command' must be a non-empty string"
            }
        }
    }
}
